controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go

```diff
@@ -324,7 +331,69 @@ func (r *KubeadmControlPlaneReconciler) updateStatus(ctx context.Context, kcp *c
 	return nil
 }
 
-func (r *KubeadmControlPlaneReconciler) upgradeControlPlane(ctx context.Context, cluster *clusterv1.Cluster, kcp *controlplanev1.KubeadmControlPlane, ownedMachines internal.FilterableMachineCollection, requireUpgrade internal.FilterableMachineCollection) (ctrl.Result, error) {
+// updateCoreDNS updates the workload cluster's kubeadm config, coredns
+// corefile and coredns deployment to match the appropriate imageTag specified
+// in the kcp DNS.
+func (r *KubeadmControlPlaneReconciler) updateCoreDNS(
+	ctx context.Context,
+	cluster *clusterv1.Cluster,
+	kcp *controlplanev1.KubeadmControlPlane,
+) error {
+	logger := r.Log.WithValues("namespace", kcp.Namespace, "kubeadmControlPlane", kcp.Name, "cluster", cluster.Name)
+	workloadCluster, err := r.managementCluster.GetWorkloadCluster(ctx, util.ObjectKey(cluster))
+	if err != nil {
+		logger.Error(err, "failed to get remote client for workload cluster", "cluster key", util.ObjectKey(cluster))
+		return err
+	}
+
+	currentVersion, _, err := workloadCluster.GetCoreDNSInfo(ctx)
+	if err != nil {
+		return errors.Wrap(err, "failed to get coredns info")
+	}
+	if err := validateCoreDNSUpgrade(currentVersion, kcp.Spec.KubeadmConfigSpec.ClusterConfiguration.DNS.ImageTag); err != nil {
+		logger.Error(err, "failed coredns version validation, skipping coredns update")
+		return nil
+	}
+
+	if err := workloadCluster.UpdateCoreDNS(ctx, &kcp.Spec.KubeadmConfigSpec.ClusterConfiguration.DNS); err != nil {
+		return errors.Wrap(err, "failed to update coredns")
+	}
+
+	return nil
+}
+
+// validateCoreDNSUpgrade returns error if the versions don't meet
+// requirements.
+// Some of the checks come from
+// https://github.com/coredns/corefile-migration/blob/v1.0.6/migration/migrate.go#L414
+func validateCoreDNSUpgrade(fromVersion, toVersion string) error {
+	from, err := semver.Parse(fromVersion)
+	if err != nil {
+		return errors.Wrapf(err, "failed to semver parse %q", fromVersion)
+	}
+	to, err := semver.Parse(toVersion)
+	if err != nil {
+		return errors.Wrapf(err, "failed to semver parse %q", toVersion)
+	}
+	if from.Compare(to) >= 0 {
+		return fmt.Errorf("toVersion %q must be greater than fromVersion %q", to.String(), from.String())
+	}
+
+	// check if the from version is even in the list of coredns versions
+	if _, ok := migration.Versions[fromVersion]; !ok {
```

Comment:
I think we may need to recreate fromVersion from `from`, stripping off any semver build metadata (e.g. if we have 1.6.7+mybuild.1, we should just use 1.6.7 when checking supported versions. (Note, don't strip prerelease data, just build metadata.)

Reply:
Oh..good. Thanks for the reminder. @vincepri mentioned that before and it was on my TODO list.